tensorboard/uploader/uploader.py

```diff
@@ -581,6 +581,157 @@ def _create_point(self, tag_proto, event, value):
+            # because we just flushed.
+            try:
+                self._add_event_internal(run_name, event, value, metadata)
+            except _OutOfSpaceError:
+                raise RuntimeError("add_event failed despite flush")
+
+    def _add_event_internal(self, run_name, event, value, metadata):
+        run_proto = self._runs.get(run_name)
+        if run_proto is None:
+            run_proto = self._create_run(run_name)
+            self._runs[run_name] = run_proto
+        tag_proto = self._tags.get((run_name, value.tag))
+        if tag_proto is None:
+            tag_proto = self._create_tag(run_proto, value.tag, metadata)
+            self._tags[(run_name, value.tag)] = tag_proto
+        self._create_point(tag_proto, event, value)
+
+    def flush(self):
+        """Sends the active request after removing empty runs and tags.
+
+        Starts a new, empty active request.
+        """
+        request = self._request
+        _prune_empty_tags_and_runs(request)
+        if not request.runs:
+            return
+
+        self._rpc_rate_limiter.tick()
+
+        with _request_logger(request, request.runs):
+            try:
+                grpc_util.call_with_retries(self._api.WriteTensor, request)
+            except grpc.RpcError as e:
+                if e.code() == grpc.StatusCode.NOT_FOUND:
+                    raise ExperimentNotFoundError()
+                logger.error("Upload call failed with error %s", e)
+
+        self._new_request()
+
+    def _create_run(self, run_name):
+        """Adds a run to the live request, if there's space.
+
+        Args:
+          run_name: String name of the run to add.
+
+        Returns:
+          The `WriteTensorRequest.Run` that was added to `request.runs`.
+
+        Raises:
+          _OutOfSpaceError: If adding the run would exceed the remaining
+            request budget.
+        """
+        run_proto = self._request.runs.add(name=run_name)
+        self._byte_budget_manager.add_run(run_proto)
+        return run_proto
+
+    def _create_tag(self, run_proto, tag_name, metadata):
+        """Adds a tag for the given value, if there's space.
+
+        Args:
+          run_proto: `WriteTensorRequest.Run` proto to which to add a tag.
+          tag_name: String name of the tag to add (as `value.tag`).
+          metadata: TensorBoard `SummaryMetadata` proto from the first
+            occurrence of this time series.
+
+        Returns:
+          The `WriteTensorRequest.Tag` that was added to `run_proto.tags`.
+
+        Raises:
+          _OutOfSpaceError: If adding the tag would exceed the remaining
+            request budget.
+        """
+        tag_proto = run_proto.tags.add(name=tag_name)
+        tag_proto.metadata.CopyFrom(metadata)
+        self._byte_budget_manager.add_tag(tag_proto)
+        return tag_proto
+
+    def _create_point(self, tag_proto, event, value):
+        """Adds a tensor point to the given tag, if there's space.
```

Comment:
What happens if a single point (which is still a full tensor) is larger than the request limit?  It looks like that should result in an OutOfSpace event, as I don't see any support for splitting a single tensor between two uploads.  Does this need something similar to the blob chunking?  The limit sounds low enough that this might be an issue.

Reply:
Good catch, a single point larger than 128Kb would just fail the upload.  I've started an internal discussion about reasonable maximum Tensor point size and filed an internal issue to track any necessary changes.   Hopefully it's just a matter of choosing a different max chunk size for WriteTensor and configuring this code and our servers to accept it.

Reply:
+1 we can raise the server-side limit and so forth, but the possibility always remains of a very large tensor exceeding whatever limit we set.  I think we can safely issue a warning, skip that tensor, and proceed-- but doing that needs some logic so that we don't just crash with RuntimeError at line 635.

Reply:
OK to punt to followup PR, per offline convo